storage/innobase/xtrabackup/src/ds_encrypt.c

```diff
@@ -552,6 +555,12 @@ encrypt_worker_thread_func(void *arg)
 		if (thd->cancelled)
 			break;
 
+		assert(gcry_md_get_algo_dlen(XB_CRYPT_HASH) ==
+		       XB_CRYPT_HASH_LEN);
+
+		memcpy(thd->to, thd->from, thd->from_len);
```

Comment:
this is an interesting part actually. I want to encrypt buffer pointed by `thd->from` and its hash and put the result into `thd->to`. I have two options:
1. allocate `hash_buffer`, put hash there, then invoke `gcry_cipher_encrypt(thd->from, thd->to); gcry_cipher_encrypt(hash_buffer, thd->to + thd->from_len)`. I.e. encrypt the buffer and the hash separately (it is streaming mode, so we are good) and put the result into the `thd->to`.
2. copy `thd->from` into `thd->to`, append hash to `thd->to`, encrypt entire chunk with single `gcry_cipher_encrypt(thd->to, thd->to)`. Libgrypt can work with the buffer inplace.

The caveat is that option 1 doesn't work for me in multithreaded environments in CentOS 5 and 6 (sigh), so I went with option 2.

As for plaintext chunks left over in the memory... Yes, we don't scrape the buffer, but it will be overwritten with the next chunk, and so on. At the moment when `xtrabackup` exists, all its memory will be given back to the system and it's kernels trouble to fill them with random garbage if needed. Isn't it?